File: src/ell/util/verbosity.py
import sys
import hashlib
import shutil
import textwrap
from typing import Dict, Tuple, List, Any, Optional
from colorama import Fore, Style, init
from ell.types import Message
from ell.configurator import config
import logging
from functools import lru_cache

from ell.types.message import LMP

# Initialize colorama
init(autoreset=True)

# Define colors and styles
ELL_COLORS = {k: v for k, v in vars(Fore).items() if k not in ['RESET', 'BLACK', 'LIGHTBLACK_EX']}
BOLD = Style.BRIGHT
UNDERLINE = '\033[4m'
RESET = Style.RESET_ALL
SYSTEM_COLOR = Fore.CYAN
USER_COLOR = Fore.GREEN
ASSISTANT_COLOR = Fore.YELLOW
PIPE_COLOR = Fore.BLUE

# Set up logging
# logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

@lru_cache(maxsize=128)
def compute_color(invoking_lmp: LMP) -> str:
    """Compute and cache a consistent color for a given LMP."""
    name_hash = hashlib.md5(invoking_lmp.__name__.encode()).hexdigest()
    color_index = int(name_hash, 16) % len(ELL_COLORS)
    return list(ELL_COLORS.values())[color_index]

def format_arg(arg: Any, max_length: int = 8) -> str:
    """Format an argument for display with customizable max length."""
    str_arg = str(arg)
    return f"{Fore.MAGENTA}{str_arg[:max_length]}..{Style.RESET_ALL}" if len(str_arg) > max_length else f"{Fore.MAGENTA}{str_arg}{Style.RESET_ALL}"

def format_kwarg(key: str, value: Any, max_length: int = 8) -> str:
    """Format a keyword argument for display with customizable max length."""
    return f"{Style.DIM}{key}{Style.RESET_ALL}={Fore.MAGENTA}{str(value)[:max_length]}..{Style.RESET_ALL}"

def get_terminal_width() -> int:
    """Get the terminal width, defaulting to 80 if it can't be determined."""
    try:
        return shutil.get_terminal_size((80, 20)).columns
    except Exception:
        logger.warning("Unable to determine terminal size. Defaulting to 80 columns.")
        return 80

def wrap_text_with_prefix(text: str, width: int, prefix: str, subsequent_prefix: str, text_color: str) -> List[str]:
    """Wrap text while preserving the prefix and color for each line."""
    paragraphs = text.split('\n')
    wrapped_paragraphs = [textwrap.wrap(p, width=width - len(prefix)) for p in paragraphs]
    wrapped_lines = [line for paragraph in wrapped_paragraphs for line in paragraph]
    result = [f"{prefix}{RESET}{wrapped_lines[0]}{RESET}" if wrapped_lines else f"{prefix}{text_color}{RESET}"]
    result.extend([f"{subsequent_prefix}{RESET}{line}{RESET}" for line in wrapped_lines[1:]])
    return result

def print_wrapped_messages(messages: List[Message], max_role_length: int, color: str, wrap_width: Optional[int] = None):
    """Print wrapped messages with proper indentation, customizable wrap width, and consistent ASCII piping."""
    terminal_width = get_terminal_width()
    prefix = f"{PIPE_COLOR}│   "
    role_prefix = ' ' * (max_role_length + 2)
    subsequent_prefix = f"{PIPE_COLOR}│   {role_prefix}"
    wrapping_width = wrap_width or (terminal_width - len(prefix))

    for i, message in enumerate(messages):
        role = message.role
        text = message.content[0].text or "" # TODO: message repr
        role_color = SYSTEM_COLOR if role == "system" else USER_COLOR if role == "user" else ASSISTANT_COLOR
        
        role_line = f"{prefix}{role_color}{role.rjust(max_role_length)}: {RESET}"
        wrapped_lines = wrap_text_with_prefix(text, wrapping_width - len(role_prefix), '', subsequent_prefix, role_color)
        
        print(f"{role_line}{wrapped_lines[0]}")
        for line in wrapped_lines[1:]:
            print(line)
        
        if i < len(messages) - 1:
            print(f"{PIPE_COLOR}│{RESET}")

def model_usage_logger_pre(
    invoking_lmp: LMP,
    lmp_args: Tuple,
    lmp_kwargs: Dict,
    lmp_hash: str,
    messages: List[Message],
    color: str = "",
    arg_max_length: int = 8
):
    """Log model usage before execution with customizable argument display length and ASCII box."""
    color = color or compute_color(invoking_lmp)
    formatted_args = [format_arg(arg, arg_max_length) for arg in lmp_args]
    formatted_kwargs = [format_kwarg(key, lmp_kwargs[key], arg_max_length) for key in lmp_kwargs]
    formatted_params = ', '.join(formatted_args + formatted_kwargs)
    
    terminal_width = get_terminal_width()
    
    logger.info(f"Invoking LMP: {invoking_lmp.__name__} (hash: {lmp_hash[:8]})")
    
    print(f"{PIPE_COLOR}╔{'═' * (terminal_width - 2)}╗{RESET}")
    print(f"{PIPE_COLOR}║ {color}{BOLD}{UNDERLINE}{invoking_lmp.__name__}{RESET}{color}({formatted_params}){RESET}")
    print(f"{PIPE_COLOR}╠{'═' * (terminal_width - 2)}╣{RESET}")
    print(f"{PIPE_COLOR}║ {BOLD}Prompt:{RESET}")
    print(f"{PIPE_COLOR}╟{'─' * (terminal_width - 2)}╢{RESET}")

    max_role_length = max(len("assistant"), max(len(message.role) for message in messages))
    print_wrapped_messages(messages, max_role_length, color)

def model_usage_logger_post_start(color: str = "", n: int = 1):
    """Log the start of model output with ASCII box."""
    terminal_width = get_terminal_width()
    print(f"{PIPE_COLOR}╟{'─' * (terminal_width - 2)}╢{RESET}")
    print(f"{PIPE_COLOR}║ {BOLD}Output{f'[0 of {n}]' if n > 1 else ''}:{RESET}")
    print(f"{PIPE_COLOR}╟{'─' * (terminal_width - 2)}╢{RESET}")
    print(f"{PIPE_COLOR}│   {ASSISTANT_COLOR}assistant: {RESET}", end='')
    sys.stdout.flush()

from contextlib import contextmanager

@contextmanager
def model_usage_logger_post_intermediate(color: str = "", n: int = 1):
    """Context manager to log intermediate model output without wrapping, only indenting if necessary."""
    terminal_width = get_terminal_width()
    prefix = f"{PIPE_COLOR}│   "
    subsequent_prefix = f"{PIPE_COLOR}│   {' ' * (len('assistant: '))}"
    chars_printed = len(subsequent_prefix)

    def log_stream_chunk(stream_chunk: str, is_refusal: bool = False):
        nonlocal chars_printed
        if stream_chunk:
            lines = stream_chunk.split('\n')
            for i, line in enumerate(lines):
                if chars_printed + len(line) > terminal_width - 6:
                    print()
                    if i == 0:
                        print(subsequent_prefix, end='')
                        chars_printed = len(prefix)
                    else:
                        print(subsequent_prefix, end='')
                        chars_printed = len(subsequent_prefix)
                    print(line.lstrip(), end='')
                else:
                    print(line, end='')
                chars_printed += len(line)

                if i < len(lines) - 1:
                    print()
                    print(subsequent_prefix, end='')
                    chars_printed = len(subsequent_prefix)  # Reset for new line
            sys.stdout.flush()

    try:
        yield log_stream_chunk
    finally:
        pass
def model_usage_logger_post_end():
    """Log the end of model output with ASCII box closure."""
    terminal_width = get_terminal_width()
    print(f"\n{PIPE_COLOR}╚{'═' * (terminal_width - 2)}╝{RESET}")

def set_log_level(level: str):
    """Set the logging level."""
    numeric_level = getattr(logging, level.upper(), None)
    if not isinstance(numeric_level, int):
        raise ValueError(f'Invalid log level: {level}')
    logger.setLevel(numeric_level)

# Example usage
# set_log_level('DEBUG')

File: typescript/src/verbosity.ts
import { Message, LMP } from './types'
import { config } from './configurator'
import { ELL_VERSION } from './_version'
import { createHash } from 'crypto'
import * as logging from '../_logger'

const logger = logging.getLogger('ell.verbosity')

function memoize<T extends (...args: any[]) => any>(fn: T, maxSize: number = 128): T {
  const cache = new Map<string, ReturnType<T>>()
  return function (this: any, ...args: Parameters<T>): ReturnType<T> {
    const key = JSON.stringify(args)
    if (cache.has(key)) {
      return cache.get(key)!
    }
    const result = fn.apply(this, args)
    if (cache.size >= maxSize) {
      cache.delete(cache.keys().next().value)
    }
    cache.set(key, result)
    return result
  } as T
}

// Colors and styles (using ANSI escape codes)
const ELL_COLORS = {
  RED: '\x1b[31m',
  GREEN: '\x1b[32m',
  YELLOW: '\x1b[33m',
  BLUE: '\x1b[34m',
  MAGENTA: '\x1b[35m',
  CYAN: '\x1b[36m',
  WHITE: '\x1b[37m',
} as const

const BOLD = '\x1b[1m'
const UNDERLINE = '\x1b[4m'
const RESET = '\x1b[0m'
const SYSTEM_COLOR = ELL_COLORS.CYAN
const USER_COLOR = ELL_COLORS.GREEN
const ASSISTANT_COLOR = ELL_COLORS.YELLOW
const PIPE_COLOR = ELL_COLORS.BLUE

let hasLoggedVersionStatement = false
const print = (s: string) => process.stdout.write(s + '\n')

async function checkVersionAndLog(): Promise<void> {
  if (!hasLoggedVersionStatement) {
    return
    try {
      const latestVersion = await fetch('https://docs.ell.so/_static/ell_version.txt')
        .then((r) => r.text())
        .then((text) => text.trim())
      if (latestVersion !== ELL_VERSION) {
        print(`${ELL_COLORS.YELLOW}╔═════════════════════════════════════════════════════════════════╗`)
        print(
          `${ELL_COLORS.YELLOW}║ ${ELL_COLORS.GREEN}A new version of ELL is available: ${
            ELL_COLORS.CYAN
          }${latestVersion.padEnd(29)}${ELL_COLORS.YELLOW}║`
        )
        print(
          `${ELL_COLORS.YELLOW}║ ${ELL_COLORS.GREEN}You can update by running:${ELL_COLORS.YELLOW}                                      ║`
        )
        print(
          `${ELL_COLORS.YELLOW}║ ${ELL_COLORS.CYAN}npm install ell-ai${ELL_COLORS.YELLOW}                                           ║`
        )
        print(`${ELL_COLORS.YELLOW}╚═════════════════════════════════════════════════════════════════╝${RESET}`)
      }
    } catch (error) {
      // Silently handle any network-related errors
    }
    hasLoggedVersionStatement = true
  }
}
function formatArg(arg: any, maxLength: number = 8): string {
  const strArg = String(arg)
  return strArg.length > maxLength
    ? `${ELL_COLORS.MAGENTA}${strArg.slice(0, maxLength)}..${RESET}`
    : `${ELL_COLORS.MAGENTA}${strArg}${RESET}`
}

function formatKwarg(key: string, value: any, maxLength: number = 8): string {
  const strValue = String(value)
  return `${RESET}${key}${RESET}=${ELL_COLORS.MAGENTA}${strValue.slice(
    0,
    maxLength
  )}${strValue.length > maxLength ? '..' : ''}${RESET}`
}
const computeColor = memoize((invokingLmp: LMP): string => {
  const nameHash = createHash('md5').update(invokingLmp.name).digest('hex')
  const colorIndex = parseInt(nameHash, 16) % Object.keys(ELL_COLORS).length
  return Object.values(ELL_COLORS)[colorIndex]
}, 128)

function wrapTextWithPrefix(
  text: string,
  width: number,
  prefix: string,
  subsequentPrefix: string,
  textColor: string
): string[] {
  const paragraphs = text.split('\n')
  const wrappedParagraphs = paragraphs.map(
    (p) => p.match(new RegExp(`.{1,${width - prefix.length}}(\\s+|$)`, 'g')) || [p]
  )
  const wrappedLines = wrappedParagraphs.flat()

  const result: string[] = []
  if (wrappedLines.length > 0) {
    result.push(`${prefix}${textColor}${wrappedLines[0]}${RESET}`)
    result.push(...wrappedLines.slice(1).map((line) => `${subsequentPrefix}${textColor}${line}${RESET}`))
  } else {
    result.push(`${prefix}${textColor}${RESET}`)
  }

  return result
}

function printWrappedMessages(messages: Message[], maxRoleLength: number, color: string, wrapWidth?: number): void {
  const terminalWidth = process.stdout.columns || 80
  const prefix = `${PIPE_COLOR}│   `
  const rolePrefix = ' '.repeat(maxRoleLength + 2)
  const subsequentPrefix = `${PIPE_COLOR}│   ${rolePrefix}`
  const wrappingWidth = wrapWidth || terminalWidth - prefix.length

  messages.forEach((message, i) => {
    const { role, text } = message
    const roleColor = role === 'system' ? SYSTEM_COLOR : role === 'user' ? USER_COLOR : ASSISTANT_COLOR

    const roleLine = `${prefix}${roleColor}${role.padStart(maxRoleLength)}: ${RESET}`
    const wrappedLines = wrapTextWithPrefix(
      text || '',
      wrappingWidth - rolePrefix.length,
      '',
      subsequentPrefix,
      roleColor
    )

    print(`${roleLine}${wrappedLines[0]}`)
    wrappedLines.slice(1).forEach((line) => print(line))

    if (i < messages.length - 1) {
      print(`${PIPE_COLOR}│${RESET}`)
    }
  })
}

export function modelUsageLoggerPre(
  invokingLmp: LMP,
  lmpArgs: any[],
  lmpKwargs: Record<string, any>,
  lmpHash: string,
  messages: Message[],
  color: string = '',
  argMaxLength: number = 8
): void {
  color = color || computeColor(invokingLmp)
  const formattedArgs = lmpArgs.map((arg) => formatArg(arg, argMaxLength))
  const formattedKwargs = Object.entries(lmpKwargs).map(([key, value]) => formatKwarg(key, value, argMaxLength))
  const formattedParams = [...formattedArgs, ...formattedKwargs].join(', ')

  checkVersionAndLog()

  const terminalWidth = process.stdout.columns || 80

  logger.info(`Invoking LMP: ${invokingLmp.name} (hash: ${lmpHash.slice(0, 8)})`)

  print(`${PIPE_COLOR}╔${'═'.repeat(terminalWidth - 2)}╗${RESET}`)
  print(`${PIPE_COLOR}║ ${color}${BOLD}${UNDERLINE}${invokingLmp.name}${RESET}${color}(${formattedParams})${RESET}`)
  print(`${PIPE_COLOR}╠${'═'.repeat(terminalWidth - 2)}╣${RESET}`)
  print(`${PIPE_COLOR}║ ${BOLD}Prompt:${RESET}`)
  print(`${PIPE_COLOR}╟${'─'.repeat(terminalWidth - 2)}╢${RESET}`)

  const maxRoleLength = Math.max('assistant'.length, ...messages.map((m) => m.role.length))
  printWrappedMessages(messages, maxRoleLength, color)
}

export function modelUsageLoggerPostStart(color: string = '', n: number = 1): void {
  const terminalWidth = process.stdout.columns || 80
  print(`${PIPE_COLOR}╟${'─'.repeat(terminalWidth - 2)}╢${RESET}`)
  print(`${PIPE_COLOR}║ ${BOLD}Output${n > 1 ? `[0 of ${n}]` : ''}:${RESET}`)
  print(`${PIPE_COLOR}╟${'─'.repeat(terminalWidth - 2)}╢${RESET}`)
  process.stdout.write(`${PIPE_COLOR}│   ${ASSISTANT_COLOR}assistant: ${RESET}`)
}

export function modelUsageLoggerPostIntermediate(
  color: string = '',
  n: number = 1
): (streamChunk: string, isRefusal?: boolean) => void {
  const terminalWidth = process.stdout.columns || 80
  const prefix = `${PIPE_COLOR}│   `
  const subsequentPrefix = `${PIPE_COLOR}│   ${''.padEnd('assistant: '.length)}`
  let charsPrinted = subsequentPrefix.length

  return function logStreamChunk(streamChunk: string, isRefusal: boolean = false): void {
    if (streamChunk) {
      const lines = streamChunk.split('\n')
      lines.forEach((line, i) => {
        if (charsPrinted + line.length > terminalWidth - 6) {
          process.stdout.write('\n')
          if (i === 0) {
            process.stdout.write(subsequentPrefix)
            charsPrinted = prefix.length
          } else {
            process.stdout.write(subsequentPrefix)
            charsPrinted = subsequentPrefix.length
          }
          process.stdout.write(line.trimStart())
        } else {
          process.stdout.write(line)
        }
        charsPrinted += line.length

        if (i < lines.length - 1) {
          process.stdout.write('\n')
          process.stdout.write(subsequentPrefix)
          charsPrinted = subsequentPrefix.length
        }
      })
    }
  }
}

export function modelUsageLoggerPostEnd(): void {
  const terminalWidth = process.stdout.columns || 80
  process.stdout.write(`\n${PIPE_COLOR}╚${'═'.repeat(terminalWidth - 2)}╝${RESET}`)
}

export function setLogLevel(level: string): void {
  // Implement logging level setting if needed
  console.log(`Log level set to: ${level}`)
}
